File: baekjoon/graph/java/BOJ21278.java
import java.io.*;
import java.util.*;

class BOJ21278 {
    static final int INF = 1000000000;
    static int N, M;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        int[][] dist = new int[N + 1][N + 1];

        for (int i=1;i<=N;i++) {
            for (int j=1;j<=N;j++) {
                if (i == j) continue;
                dist[i][j] = INF;
            }
        }

        for (int i=0;i<M;i++) {
            st = new StringTokenizer(br.readLine());

            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());

            dist[u][v] = 1;
            dist[v][u] = 1;
        }

        for (int k=1;k<=N;k++){
            for (int i=1;i<=N;i++) {
                for (int j=1;j<=N;j++) {
                    if (i == j) continue;
                    dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }


        int ans = INF;
        int c1 = 0, c2 = 0;
        for (int i=1;i<=N;i++) {
            for (int j=i+1;j<=N;j++) {
                int current = 0;
                for (int k=1;k<=N;k++){
                    if (i == k || j == k) continue;
                    
                    current += Math.min(dist[i][k], dist[j][k]);
                }

                if (current < ans) {
                    c1 = i;
                    c2 = j;
                    ans = current;
                }
            }
        }

        System.out.println(c1 +" " + c2 +" "+ (ans*2));

    }
}

File: 0_1.baekjoon/graph/python/BOJ21278.py
from sys import stdin
from collections import defaultdict

readline = stdin.readline

INF = float('inf')
N, M = map(int, readline().split())
dist = [[INF for _ in range(N + 1)] for _ in range(N + 1)]

def solve():

    for i in range(N + 1):
        dist[i][i] = 0

    for _ in range(M):
        u, v = map(int, readline().split())
        dist[u][v] = 1
        dist[v][u] = 1

    for k in range(1, N + 1):
        for i in range(1, N + 1):
            for j in range(1, N + 1):
                if i == j: continue

                if dist[i][j] > dist[i][k] + dist[k][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]


    ans = [-1, -1, INF]
    for i in range(1, N):
        for j in range(i + 1, N + 1):

            if ans[0] == -1:
                ans = [i, j, get_dist(i, j, dist)]
                continue
            
            current = get_dist(i, j, dist)

            if ans[-1] > current:
                ans = [i, j, current]

    print(*ans)

def get_dist(u, v, dist):

    result = 0

    for i in range(1, N + 1):
        result += min(dist[u][i], dist[v][i])

    return result * 2

solve()
